src/back/space/src/api/invitation.cpp

```diff
@@ -0,0 +1,266 @@
+#include <boost/lexical_cast.hpp>
+
+#include "invitation.hpp"
+#include "../service/service.hpp"
+#include "../../../shared/headers.hpp"
+#include "../../../shared/errors.hpp"
+#include "../../../shared/paging.hpp"
+#include "../../../shared/paging.сpp"
+#include "../model/role.hpp"
+#include "../model/invitation_serialize.hpp"
+
+namespace svetit::space::handlers {
+
+Invitation::Invitation(
+	const components::ComponentConfig& conf,
+	const components::ComponentContext& ctx)
+	: server::handlers::HttpHandlerJsonBase{conf, ctx}
+	, _s{ctx.FindComponent<Service>()}
+{}
+
+formats::json::Value Invitation::GetList(
+	const server::http::HttpRequest& req,
+	const formats::json::Value& body) const
+{
+	formats::json::ValueBuilder res;
+
+	try {
+		auto paging = parsePaging(req);
+		res["list"] = _s.GetInvitationList(paging.start, paging.limit);
+		res["total"] = _s.GetInvitationsCount();
+	}
+	catch(const errors::BadRequest& e) {
+		res["err"] = e.what();
+		req.SetResponseStatus(server::http::HttpStatus::kBadRequest);
+		return res.ExtractValue();
+	}
+	catch(const std::exception& e) {
+		LOG_WARNING() << "Fail to get invitations list: " << e.what();
+		req.SetResponseStatus(server::http::HttpStatus::kInternalServerError);
+		res["err"] = "Fail to get invitations list";
+	}
+
+	return res.ExtractValue();
+}
+
+formats::json::Value Invitation::Post(
+	const server::http::HttpRequest& req,
+	const formats::json::Value& body) const
+{
+	formats::json::ValueBuilder res;
+
+	try {
+		const auto& creatorId = req.GetHeader(headers::kUserId);
+		if (creatorId.empty())
+			throw errors::Unauthorized{};
+
+		if (req.HasArg("link")) {
+			const auto link = req.GetArg("link");
+
+			if (link.empty())
+				throw errors::BadRequest{"Empty linkId"};
+
+			if (!_s.InviteByLink(creatorId, link))
+				throw errors::BadRequest{"Failed invite by link"};
+
+			req.SetResponseStatus(server::http::HttpStatus::kCreated);
+			return res.ExtractValue();
+		}
+
+		auto invitation = body.As<model::SpaceInvitation>();
+
+		if (invitation.spaceId.is_nil() || invitation.userId.empty())
+			throw errors::BadRequest{"Params must be set"};
+
+		if (!_s.ValidateRole(invitation.role))
+			throw errors::BadRequest{"Wrong role"};
+
+		if (!_s.Invite(creatorId, invitation.spaceId, invitation.userId, invitation.role))
+			throw errors::BadRequest{"Failed to invite"};
```

Comment:
У нас сейчас в данной функции _s.Invite(...) закомментирован функционал (бизнес-логика проверки на возможность создания приглашение), который будет выбрасывать исключения. 

```
bool Service::Invite(const std::string& creatorId, const boost::uuids::uuid& spaceUuid, const std::string& userId, const Role::Type& role) {

	// bool isPossibleToInvite = false;

	// if (!_repo.Space().IsRequestsAllowed(spaceUuid)) {
	// 	if (_repo.SpaceUser().IsAdminInSpace(creatorUuid, spaceUuid)) {
	// 		if (!_repo.SpaceUser().IsInSpace(userUuid, spaceUuid)) {
	// 			isPossibleToInvite = true;
	// 		}
	// 	}
	// } else {
	// 	if (!_repo.SpaceUser().IsInSpace(creatorUuid, spaceUuid)) {
	// 		if (creatorUuid == userUuid) {
	// 			if (!_repo.SpaceInvitation().IsExist(spaceUuid, creatorUuid, creatorUuid)) {
	// 				isPossibleToInvite = true;
	// 			}
	// 		}
	// 	} else {
	// 		if (_repo.SpaceUser().IsAdminInSpace(creatorUuid, spaceUuid)){
	// 			if (!_repo.SpaceUser().IsInSpace(userUuid, spaceUuid)) {
	// 				isPossibleToInvite = true;
	// 			}
	// 		}
	// 	}
	// }

	// todo - true harcoded by now because isPossibleToInvite logic is not ready yet
	if (true) {
		// todo - is need to check that space and users with spaceUuid, userId, creatorId exist?
		const auto p1 = std::chrono::system_clock::now();
		const auto now = std::chrono::duration_cast<std::chrono::seconds>(p1.time_since_epoch()).count();
		_repo.SpaceInvitation().Insert(spaceUuid, userId, role, creatorId, now);
	} else {
		return false;
	}

	return true;
}
```

Мы планировали его обсудить на предмет параллельного выполнения (+ переписать на SQL по максимуму). Пока мы не обсудили, в функции ничего не проверяется и false даже не может быть возвращен, соответственно разбивать логику на отдельные функции, как ты написал, пока нельзя. Давай обсудим эту логику?

Reply:
Готово.